### PR TITLE
Removing eig solver

### PR DESCRIPTION
**Pull request by rosecers**
Removing the slower, code-streamlined eigensolver from pcovr. Previously I had been a bit weighed down by the sklearn eigendecomposition code (you'll notice the extent in the code), but this is the last bit to change before starting a PR with scikit-learn. @agoscinski , can you review?

Note a few things:
the pieces in `from_sklearn` are not permanent
I added tests to appease the codecov gods, and they are minute

### CODE REVIEW

**Review by agoscinski — COMMENTED** [review]
The user input for PCovR is well tested, I thought about adding an identity test: For the n_component  = n_features case the transformation should just a unitary transformation without information loss.
```py
  pcovr = self.model(mixing=1, n_components=self.X.shape[1], tol=1e-12)
  T = pcovr.fit_transform(self.X, self.Y)
  pcovr.inverse_transform(T)
  print(np.linalg.norm(self.X-T))
  self.assertTrue(np.allclose(self.X, T))
```
but it fails with a huge error, do you have an idea why? This also does not work on the main branch.

**Comment by rosecers** [comment]
> The user input for PCovR is well tested, I thought about adding an identity test: For the n_component = n_features case the transformation should just a unitary transformation without information loss.
> 
> ```python
>   pcovr = self.model(mixing=1, n_components=self.X.shape[1], tol=1e-12)
>   T = pcovr.fit_transform(self.X, self.Y)
>   pcovr.inverse_transform(T)
>   print(np.linalg.norm(self.X-T))
>   self.assertTrue(np.allclose(self.X, T))
> ```
> 
> but it fails with a huge error, do you have an idea why? This also does not work on the main branch.

Even with a n_components=X.shape[1] transformation, `T` will not equal `X`, because T will, by definition, be trying to pack as much variance into the first few components, so it will be decidedly different. If you wanted to check that T contains the full variance of X, you can compare their variances as a test.

File: skcosmo/pcovr/pcovr.py [hunk 1]
```diff
@@ -1,17 +1,24 @@
...
+from sklearn.utils.validation import check_is_fitted
 from sklearn.utils import check_array
-from sklearn.decomposition._base import _BasePCA
-from sklearn.linear_model._base import LinearModel
-
-from skcosmo.utils import eig_solver
+from sklearn.utils import check_random_state
+from sklearn.utils.extmath import randomized_svd, svd_flip
+from sklearn.utils.extmath import stable_cumsum
+from .from_sklearn import _infer_dimension, _init_arpack_v0, _BasePCA, LinearModel
```
[the remark this answers is not in the document]
Comment (rosecers) [reply skcosmo/pcovr/pcovr.py:16]: I've had some issues using it in this manner on my remote workstation, but not my laptop. The idea is that this is a temporary band-aid, I just have to figure out what's up to make sure it's compatible for whatever computer.

File: skcosmo/pcovr/pcovr.py [hunk 2]
```diff
@@ -39,16 +46,37 @@ def pcovr_covariance(mixing, X, Y, rcond=1e-12, return_isqrt=False):
...
+        n_components = min(X.shape)
+
     cov = X.T @ X
 
     if mixing < 1 or return_isqrt:
-        # Do not try to approximate C_inv, it will affect results
-        C_inv = np.linalg.pinv(cov, rcond=rcond)
-        C_isqrt = np.real(scipy.linalg.sqrtm(C_inv))
+
+        _, vC, UC = randomized_svd(
```
[the remark this answers is not in the document]
Comment (rosecers) [reply skcosmo/pcovr/pcovr.py:72]: So here, because we're inverting the eigenvalues, we are almost never going to be wanting the full eigendecomposition, otherwise the square root inverse will be unphysically large. From the sklearn options, this eliminates "full" as a reasonable choice. I can modify it to consider arpack as well, although I chose randomized here because it is the sklearn default for truncated SVDs. Do you think it's crucial to have both?

File: skcosmo/pcovr/pcovr.py [hunk 3]
```diff
@@ -107,13 +135,32 @@ class PCovR(_BasePCA, LinearModel):
     Parameters
     ----------
     mixing: float, defaults to 1
-        mixing parameter, as described in PCovR as :math:`{\alpha}`
+        mixing parameter, as described in PCovR as :math:`{\alpha}`, here named
+        to avoid confusion with regularization parameter `alpha`
 
-    n_components : int, float or str, default=None
+    self.n_components : int, float or str, default=None
```
[the remark this answers is not in the document]
Comment (rosecers) [reply skcosmo/pcovr/pcovr.py:141]:
```suggestion
    n_components : int, float or str, default=None
```

File: skcosmo/pcovr/pcovr.py [hunk 4]
```diff
@@ -170,7 +226,7 @@ class PCovR(_BasePCA, LinearModel):
     explained_variance_ : ndarray of shape (n_components,)
         The amount of variance explained by each of the selected components.
 
-        Equal to n_components largest eigenvalues
+        Equal to self.n_components largest eigenvalues
```
[the remark this answers is not in the document]
Comment (rosecers) [reply skcosmo/pcovr/pcovr.py:229]:
```suggestion
        Equal to n_components largest eigenvalues
```

File: skcosmo/pcovr/pcovr.py [hunk 5]
```diff
@@ -485,7 +687,7 @@ def transform(self, X=None):
         return super().transform(X)
 
     def score(self, X, Y, T=None):
-        r"""Return the normalized, squared reconstruction error for X and Y,
+        r"""Return the totalreconstruction error for X and Y,
```
[the remark this answers is not in the document]
Comment (rosecers) [reply skcosmo/pcovr/pcovr.py:690]:
```suggestion
        r"""Return the total reconstruction error for X and Y,
```

File: skcosmo/pcovr/pcovr.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (agoscinski) [reply skcosmo/pcovr/pcovr.py:16]: sure

[the remark this answers is not in the document]
Comment (agoscinski) [reply skcosmo/pcovr/pcovr.py:16]: maybe add a note that this file is temporary and will be removed is future, if you feel like this could take longer

**Comment by agoscinski** [comment]
Sorry the code I have sent you is a bit wrong, I wanted to use the inverse transformation of T not T itself. This works. What do you think about adding this test for a simple check if the PCA case in PCovR works?
```py
  pcovr = self.model(mixing=1, n_components=self.X.shape[1], tol=1e-12)
  T = pcovr.fit_transform(self.X, self.Y)
  T_inv = pcovr.inverse_transform(T)
  self.assertTrue(np.allclose(self.X, T_inv))
```

There are errors in the examples which dont result in a failure of the github workflow checks, do you know why? (maybe @Luthaf can have a look) 
See output of "Run tests" tab https://github.com/cosmo-epfl/scikit-cosmo/pull/53/checks?check_run_id=1790911839

File: skcosmo/pcovr/pcovr.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (agoscinski) [reply skcosmo/pcovr/pcovr.py:72]: But couldn't you do the same argument for the svd PCovR?
Also in the PCovR class in `_fit_feature_space`, the `n_components` is not an input argument for the function `pcovr_covariance`, so the PCA svd solver would theoretically default to the full method in this case.

File: skcosmo/pcovr/pcovr.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rosecers) [reply skcosmo/pcovr/pcovr.py]: So I found the issue -- those methods only exist on scikit-learn 0.24.0, which is the newest release (my workstation was on 0.23dev). We shouldn't need the file anymore.

**Comment by rosecers** [comment]
> Sorry the code I have sent you is a bit wrong, I wanted to use the inverse transformation of T not T itself. This works. What do you think about adding this test for a simple check if the PCA case in PCovR works?
> 
> ```python
>   pcovr = self.model(mixing=1, n_components=self.X.shape[1], tol=1e-12)
>   T = pcovr.fit_transform(self.X, self.Y)
>   T_inv = pcovr.inverse_transform(T)
>   self.assertTrue(np.allclose(self.X, T_inv))
> ```
> 
> There are errors in the examples which dont result in a failure of the github workflow checks, do you know why? (maybe @Luthaf can have a look)
> See output of "Run tests" tab https://github.com/cosmo-epfl/scikit-cosmo/pull/53/checks?check_run_id=1790911839

It's a trivial add, and I don't mind. Adding now.

File: skcosmo/pcovr/pcovr.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rosecers) [reply skcosmo/pcovr/pcovr.py:76]: You want that -- the n_components used to construct Ctilde should not be the same as the n_components used for the projection. Maybe I should rename the n_components in constructing Ctilde? 

Essentially, it is incredibly costly to compute the inverse square root through usual methods (scipy.linalg.sqrtm), so an approximation of Ctilde will suffice using the eigendecomposition. _However_ this approximation shouldn't be low-rank like the PCA, but rather high enough rank to correctly construct Ctilde. I called it n_components to be consistent with the randomized_svd naming, but if it leads to confusion with the n_PCA components, it might be worth changing.

[the remark this answers is not in the document]
Comment (rosecers) [reply skcosmo/pcovr/pcovr.py:76]: I've renamed it `rank` to avoid this confusion. It should infer the rank that you assume the covariance is.

**Review by agoscinski — APPROVED** [review]
Looks good!
